proposals/001-ClusterRepresentation.md

```diff
@@ -0,0 +1,344 @@
+
+A _physical cluster_ is a model representation of a real Apache Kafka cluster.  There is always a one to one correspondance between the _physical cluster_
+and a real Apache Kafka cluster. Many physical clusters can be defined with a Kroxylicious instance. Each _physical cluster_ comprises one or more _physical brokers_.  Conceptually, the physical cluster exists on the [upstream](https://github.com/kroxylicious/design/blob/main/concepts.asciidoc#upstream-vs-downstream) (client side) of Kroxylicious.
+
+There will be a mechansim to map between virtual and physical clusters.  This will allow the building useful topologies to serve different use-cases, such as:
+
+- *one to one* - the proxying of a single physical cluster.
+- *many to one* - kroxylicious presents many virtual clusters which map to a physical cluster. This would support a multi-tenant use-case where a single physical clusters is shared by isolated tenants.
+
+## Illustration of many-to-one use-case
+
+![alternative text](https://raw.githubusercontent.com/kroxylicious/design/001-ClusterRepresentation/proposals/001/many-to-one.png)
+
+## Goals
+
+* Ability to support the *one to one* and *many to one* use-cases described above.
+* Ability to add/remove/change virtual or physical clusters at runtime, without interrupting connections established.
+
+
+## Non-Goals
+
+* There's a possiblity of *one-to-many* where a single virtual cluster maps to several physical, presenting them as if they were one. However, supporting transactions across two or more physical clusters would be difficult.  This use-case is out-of-scope.
+* It might be possible for a virtual cluster to present a subset of a physical cluster's brokers as virtual brokers by clevery exposing only those brokers that host topic partitions belonging to that virtual cluster.  This might be advantagous in the case where the physical cluster comprises a very large number of brokers.  We won't consider this use-case as part of this proposal.
+
+
+## Current situation
+
+Kroxyliciousis currently limited to exposing a single broker of a single cluster.
+
+
+## Proposal
+
+The proposal will change some existing concepts and introduce some new ones.  These are discussed first.  Then, we talk about the changes in high level responsibilities.
+
+### Key Objects
+
+
+![alternative text](http://www.plantuml.com/plantuml/proxy?cache=no&src=https://raw.github.com/kroxylicious/design/001-ClusterRepresentation/proposals/001/UML/uml.puml)
+
+
+#### Endpoints
+
+Currently `KafkaProxy` uses the `ProxyConfig` object to get a single address to bind to.  
+
+* For the non-TLS case, multiple endpoints are required to support >1 brokers. 
+* For TLS case, SNI gives a mechanism to route the traffic to >1 brokers even when a single endpoint is used, however there may be performance or operational reasons to favour the ability to define endpoints.
+
+`ProxyConfig` will be replaced by an list of _Endpoint_ objects.  
```

Comment:
I was thinking the endpoint specify if it was TLS or not. It would probably also be where we configure cipher suites etc though I haven't specified this.  If we pushed the endpoint configuration into the upstream, we'd need to resolve any conflicts. I think having named endpoints is simpler to avoid those issues.

Reply:
Yeah ok. Lets leave them separated.